### PR TITLE
Cleaned up core

### LINKED ISSUE #488: Clean up simple core
-Purne simple core so that it only has required code.
- Find the usage of `CoreInterface` and replace it with `StateRootInterface`.
- Rename `StateRootCommited` event in simple core to `StateRootAvailable`. (This event will be avaialbe in Mosaic core and also Origin block store on finalization. This is a separate issue)
- Variable `coreOpenSTRemote`, `coreChainIdOrigin` can be removed.
- Simple core must be an organised contract.
- Variable `workers` can be removed, this can be handled due to organised contract.
- Organisation address can set the Co-Simple Core contract address.
- Write/Update all the unit test cases for the modified/implemented functions.

### PR DESCRIPTION
**Pull request by schemar**
* Removed `CoreInterface` and replaced it with `StateRootInterface`
  where possible.
* Renamed `StateRootCommitted` to `StateRootAvailable`.
* Removed `coreOpenStRemote` and `coreChainIdOrigin`.
* Simple core is now `Organized`.
* Removed variable workers and used modifiers from `Organized`.
* Only organization can set the CoCore address.
* Only organization can (de)activate (co-)gateways.
* Updated tests.
* Wrote new tests for setting the CoCore address.
* Updated `Organized` to have `onlyWorker()` and `onlyOrganization()`.
  As a result, I deleted `IsWorkerInterface` and replaced it with
  `OrganizationInterface` everywhere. I also added `isWorker` and
  `isOwner` methods to the interface.

Fixes #488

### CODE REVIEW

File: contracts/lib/OrganizationInterface.sol [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+/**
+ *  @title OrganizationInterface provides methods to manage an organization.
+ */
+interface OrganizationInterface {
```
Comment (deepesh-kn) [inline contracts/lib/OrganizationInterface.sol]: We can have a slim interface for organized.

File: contracts/gateway/Core.sol [hunk 2]
```diff
@@ -119,10 +93,14 @@ contract Core is CoreInterface, StateRootInterface {
...
-        onlyWorker
+        onlyOrganization
         returns (bool /*success*/)
     {
-        require(_coCore != address(0), "Co-Core address is 0");
+        require(
+            _coCore != address(0),
+            "Co-Core address must not be 0."
+        );
+
```
Comment (deepesh-kn) [inline contracts/gateway/Core.sol]: coCore should be set only once.

File: contracts/lib/OrganizationInterface.sol (hunk 1 above)
Comment (schemar) [reply contracts/lib/OrganizationInterface.sol]: Done in `IsMemberInterface`.

File: test/gateway/core.js [hunk 3]
```diff
@@ -37,30 +34,72 @@ contract('Core', function (accounts) {
...
+            worker = contractsData.worker;
+
+            coCoreAddress = accounts[8];
         });
 
-        it('has workers', async () => {
-            assert.equal(await core.workers.call(), workersContract.address);
-            let latestStateRootBlockHeight = await core.getLatestStateRootBlockHeight.call();
+        it('should allow the organization to set the address', async () => {
+            await core.setCoCoreAddress(coCoreAddress, { from: owner });
```
Comment (deepesh-kn) [inline test/gateway/core.js]: There is not assert for this test if it is successful and actually set the coCore address.

File: test/gateway/core.js [hunk 4]
```diff
@@ -70,12 +109,12 @@ contract('Core', function (accounts) {
         });
 
         it('should be able to commit state root and getStateRoot for given block height', async () => {
-            let response = await core.commitStateRoot(blockHeight, stateRoot, {from: worker})
-            ;
+            let response = await core.commitStateRoot(blockHeight, stateRoot, { from: worker })
```
Comment (deepesh-kn) [inline test/gateway/core.js]: we should also check for the return type using `core.commitStateRoot.call`

File: test/gateway/core.js [hunk 5]
```diff
@@ -85,19 +124,19 @@ contract('Core', function (accounts) {
         });
 
         it('should not be able to commit state root of block height which is equal to latest block height', async () => {
-            await utils.expectThrow(core.commitStateRoot(blockHeight, stateRoot, {from: worker}));
+            await utils.expectThrow(core.commitStateRoot(blockHeight, stateRoot, { from: worker }));
```
Comment (deepesh-kn) [inline test/gateway/core.js]: In the tests all the `it` looks interdependent. 
Should we consider to split this file into the function-specific test? (In this ticket or create a new ticket? Please suggest.)

File: test/gateway/core.js [hunk 6]
```diff
@@ -37,30 +34,72 @@ contract('Core', function (accounts) {
             blockHeight = new BN(5);
             contractsData = await coreUtils.deployCore(artifacts, accounts);
             core = contractsData.core;
-            workersContract = contractsData.workersContract;
             worker = contractsData.worker;
             registrar = contractsData.registrar;
```
Comment (deepesh-kn) [inline test/gateway/core.js]: We don't use `registrar` anymore.

File: test/gateway/core_utils.js [hunk 7]
```diff
@@ -19,45 +19,29 @@
...
-    const core = await Core.new(chainIdOrigin, chainIdRemote, 0, proof.account.stateRoot, workers.address, {from: accounts[0]});
+    const organization = await Organization.new(organizationOwner, worker);
+    const core = await Core.new(chainIdRemote, 0, proof.account.stateRoot, organization.address, { from: accounts[0] });
     return {
         core: core,
-        workersContract: workers,
-        worker: worker1,
+        owner: organizationOwner,
+        worker: worker,
         registrar: registrar,
```
Comment (deepesh-kn) [inline test/gateway/core_utils.js]: We don't use `registrar` anymore.

File: test/gateway/core_utils.js [hunk 8]
```diff
@@ -68,7 +52,7 @@ module.exports.checkOpenSTProvenEvent = (event, _blockHeight, _storageRoot, wasA
     assert.equal(event["wasAlreadyProved"], wasAlreadyProved);
 };
 
-module.exports.checkStateRootCommittedEvent = (event, _blockHeight, _stateRoot) => {
+module.exports.checkStateRootAvailableEvent = (event, _blockHeight, _stateRoot) => {
```
Comment (deepesh-kn) [inline test/gateway/core_utils.js]: L48: We are not using `checkOpenSTProvenEvent` anymore, we can delete it.

File: contracts/gateway/Core.sol [hunk 9]
```diff
@@ -21,93 +21,71 @@ pragma solidity ^0.5.0;
...
+import "../lib/IsMemberInterface.sol";
+import "../lib/MerklePatriciaProof.sol";
+import "../lib/Organized.sol";
 import "../lib//RLP.sol";
 import "../lib/SafeMath.sol";
-import "../StateRootInterface.sol";
 
 /**
- *  @title Core contract which implements CoreInterface.
+ *  @title Core contract which implements StateRootInterface.
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/gateway/Core.sol]: done

File: contracts/gateway/Core.sol [hunk 10]
```diff
@@ -21,93 +21,71 @@ pragma solidity ^0.5.0;
...
  *
  *  @notice Core is a minimal stub that will become the anchoring and consensus point for
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/gateway/Core.sol]: done

File: contracts/gateway/Core.sol [hunk 11]
```diff
@@ -21,93 +21,71 @@ pragma solidity ^0.5.0;
...
  */
-contract Core is CoreInterface, StateRootInterface {
+contract Core is StateRootInterface, Organized {
     using SafeMath for uint256;
 
+
     /** Events */
 
-    event StateRootCommitted(uint256 blockHeight, bytes32 stateRoot);
+    event StateRootAvailable(uint256 blockHeight, bytes32 stateRoot);
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/gateway/Core.sol]: OK then I will update it. But we should overthink this. In web3 1.0, you can access the event parameters by their names, and I am not sure we want users in the JS layer to access these with a starting underscore.

/cc @benjaminbollen @deepesh-kn @gulshanvasnani @sarvesh-ost

File: contracts/gateway/Core.sol [hunk 12]
```diff
@@ -21,93 +21,71 @@ pragma solidity ^0.5.0;
...
     mapping (uint256 /* block height */ => bytes32) private stateRoots;
 
-    /** chainIdOrigin is the origin chain id where core contract is deployed.  */
-    uint256 public coreChainIdOrigin;
     /** chainIdRemote is the remote chain id where core contract is deployed. */
     uint256 private coreChainIdRemote;
-    /** It is the address of the openSTUtility/openSTValue contract on the remote chain. */
-    address private coreOpenSTRemote;
     /** Latest block height of block for which state root was committed. */
     uint256 private latestStateRootBlockHeight;
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/gateway/Core.sol]: done

File: contracts/gateway/Core.sol [hunk 13]
```diff
@@ -21,93 +21,71 @@ pragma solidity ^0.5.0;
...
-    }
-
-    /*  Public functions */
+    /*  Constructor */
 
     /**
      *  @notice Contract constructor.
      *
-     *  @param _chainIdOrigin Chain id where current core contract is deployed since core contract can be deployed on remote chain also.
      *  @param _chainIdRemote If current chain is value then _chainIdRemote is chain id of utility chain.
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/gateway/Core.sol]: updated

File: contracts/gateway/Core.sol [hunk 14]
```diff
@@ -119,28 +97,24 @@ contract Core is CoreInterface, StateRootInterface {
      */
     function setCoCoreAddress(address _coCore)
         external
-        onlyWorker
-        returns (bool /*success*/)
+        onlyOrganization
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/gateway/Core.sol]: What do you mean? Can you please be more specific?

File: contracts/gateway/Core.sol [hunk 15]
```diff
@@ -171,8 +145,6 @@ contract Core is CoreInterface, StateRootInterface {
         return latestStateRootBlockHeight;
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/gateway/Core.sol]: fixed

File: contracts/gateway/Core.sol [hunk 16]
```diff
@@ -200,8 +172,24 @@ contract Core is CoreInterface, StateRootInterface {
         stateRoots[_blockHeight] = _stateRoot;
         latestStateRootBlockHeight = _blockHeight;
 
-        emit StateRootCommitted(_blockHeight, _stateRoot);
+        emit StateRootAvailable(_blockHeight, _stateRoot);
 
         return _stateRoot;
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/gateway/Core.sol]: agree. updated.

File: contracts/gateway/Core.sol [hunk 17]
```diff
@@ -200,8 +172,24 @@ contract Core is CoreInterface, StateRootInterface {
...
+
+    /**
+     *  @notice Public view function chainIdRemote.
+     *
+     *  @return uint256 coreChainIdRemote.
+     */
+    function chainIdRemote()
+        public
+        view
+        returns (uint256 /* chainIdRemote */)
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/gateway/Core.sol]: updated

File: test/gateway/core.js (hunk 6 above)
Comment (schemar) [reply test/gateway/core.js]: 👀 removed

File: test/gateway/core.js (hunk 3 above)
Comment (schemar) [reply test/gateway/core.js]: added

File: test/gateway/core.js (hunk 4 above)
Comment (schemar) [reply test/gateway/core.js]: added

File: test/gateway/core.js (hunk 5 above)
Comment (schemar) [reply test/gateway/core.js]: I would prefer a separate ticket to clean up the tests.

File: test/gateway/core_utils.js (hunk 7 above)
Comment (schemar) [reply test/gateway/core_utils.js]: removed

File: test/gateway/core_utils.js (hunk 8 above)
Comment (schemar) [reply test/gateway/core_utils.js]: 🦅 removed

File: contracts/gateway/Core.sol (hunk 14 above)
[the remark this answers is not in the document]
Comment (deepesh-kn) [reply contracts/gateway/Core.sol]: I am referring to this comment.

```
     * @notice The Co-Core address is the address of the core that is
     *          deployed on the auxiliary chain. Should only be set if this
     *          contract is deployed on the origin chain.
```    
The comment specifies that the co-core is the address is of auxiliary chain, and it should be seti only for core of origin chain.

While writing the initial code, the understanding was not clear, so the comment is incorrect.

File: test/gateway/core.js (hunk 5 above)
Comment (deepesh-kn) [reply test/gateway/core.js]: Created a ticket for this #535

File: contracts/gateway/Core.sol [hunk 18]
```diff
@@ -21,93 +21,77 @@ pragma solidity ^0.5.0;
...
 // ----------------------------------------------------------------------------
 
-import "./CoreInterface.sol";
-import "../lib/MerklePatriciaProof.sol";
+
 import "./WorkersInterface.sol";
+import "../StateRootInterface.sol";
+import "../lib/IsMemberInterface.sol";
+import "../lib/MerklePatriciaProof.sol";
+import "../lib/Organized.sol";
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/gateway/Core.sol]: @sarvesh-ost I don't understand. Can you please clarify?

File: contracts/gateway/Core.sol [hunk 19]
```diff
@@ -21,93 +21,77 @@ pragma solidity ^0.5.0;
...
     /** Storage */
 
-    mapping (uint256 /* block height */ => bytes32) private stateRoots;
+    /** Maps block heights to their respective state root. */
+    mapping (uint256 => bytes32) private stateRoots;
 
-    /** chainIdOrigin is the origin chain id where core contract is deployed.  */
-    uint256 public coreChainIdOrigin;
     /** chainIdRemote is the remote chain id where core contract is deployed. */
     uint256 private coreChainIdRemote;
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/gateway/Core.sol]: Good question. @benjaminbollen wrote this more than a year ago. I will rename it to `remateChainId`.

File: contracts/gateway/Core.sol (hunk 14 above)
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/gateway/Core.sol]: ok I removed the notion that this should only be done on the origin chain.

File: contracts/gateway/Core.sol [hunk 20]
```diff
@@ -190,18 +169,38 @@ contract Core is CoreInterface, StateRootInterface {
...
-        returns (bytes32 /* stateRoot */)
+        returns (bool success_)
     {
         // State root should be valid
         require(_stateRoot != bytes32(0), "State root is 0");
+
         // Input block height should be valid
-        require(_blockHeight > latestStateRootBlockHeight, "Given block height is lower or equal to highest committed state root block height.");
+        require(
+            _blockHeight > latestStateRootBlockHeight,
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/gateway/Core.sol]: Not sure why @abhayks1 added this originally. However, I think it does make sense with the upcoming change where we will only store the last `n` reported state roots. In that case, it looks reasonable to only allow the reporting of *newer* state roots. What do you think?

/cc @deepesh-kn @abhayks1

File: contracts/gateway/Core.sol [hunk 21]
```diff
@@ -190,18 +169,38 @@ contract Core is CoreInterface, StateRootInterface {
...
+        success_ = true;
+    }
+
+
+    /* Public functions */
+
+    /**
+     *  @notice Public view function chainIdRemote.
+     *
+     *  @return uint256 coreChainIdRemote.
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/gateway/Core.sol]: done

File: contracts/gateway/Core.sol [hunk 22]
```diff
@@ -190,18 +169,38 @@ contract Core is CoreInterface, StateRootInterface {
...
+     */
+    function chainIdRemote()
+        public
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/gateway/Core.sol]: Not sure. @benjaminbollen did this a year ago and I think it is simply outdated. Updated to be external.

File: contracts/gateway/GatewayBase.sol [hunk 23]
```diff
@@ -75,8 +74,11 @@ contract GatewayBase is Organized {
      */
     MessageBus.MessageBox messageBox;
 
-    /** address of core contract. */
-    CoreInterface public core;
+    /** Specifies if the Gateway is activated for any new process. */
+    bool public activated;
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/gateway/GatewayBase.sol]: this was removed when merging the latest `develop`.

File: contracts/test/core/MockCore.sol [hunk 24]
```diff
@@ -1,6 +1,6 @@
 pragma solidity ^0.5.0;
 
-import "../../gateway/WorkersInterface.sol";
+import "../../lib/OrganizationInterface.sol";
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/test/core/MockCore.sol]: removed 👍

File: contracts/test/core/MockCore.sol [hunk 25]
```diff
@@ -16,25 +16,23 @@ contract MockCore is Core {
     /**
      * @notice Contract constructor
      *
-     * @param _chainIdOrigin chain id where current core contract is deployed since core contract can be deployed on remote chain also
-     * @param _chainIdRemote if current chain is value then _chainIdRemote is chain id of utility chain
-     * @param _blockHeight block height at which _stateRoot needs to store
-     * @param _stateRoot state root hash of given _blockHeight
-     * @param _workers Workers contract address
+     * @param _chainIdRemote if current chain is value then _chainIdRemote is
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/test/core/MockCore.sol]: updated to match new safe core comments

File: contracts/test/core/MockCore.sol [hunk 26]
```diff
@@ -16,25 +16,23 @@ contract MockCore is Core {
...
+     *                       chain id of utility chain.
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/test/core/MockCore.sol]: updated

File: contracts/test/lib/MockOrganization.sol [hunk 27]
```diff
@@ -0,0 +1,150 @@
...
+        _proposedOwner;
+        success_ = true;
+    }
+
+    /**
+     * @notice Complete ownership transfer to proposed owner.
+     *
+     * @return success_ True on successful execution.
+     */
+    function completeOwnershipTransfer() external returns (bool success_) {
```
[the remark this answers is not in the document]
Comment (schemar) [reply contracts/test/lib/MockOrganization.sol]: we need it to satisfy the interface

File: contracts/gateway/Core.sol (hunk 18 above)
[the remark this answers is not in the document]
Comment (sarvesh-ost) [reply contracts/gateway/Core.sol]: `import "../lib//RLP.sol";` . -> `import "../lib/RLP.sol";`

**Pull request merged**